src/EditorFeatures/Core/Implementation/ForegroundNotification/ForegroundNotificationService.cs

```diff
@@ -35,7 +36,11 @@ public ForegroundNotificationService()
             _workQueue = new PriorityQueue();
             _lastProcessedTimeInMS = Environment.TickCount;
 
-            Task.Factory.SafeStartNewFromAsync(ProcessAsync, CancellationToken.None, TaskScheduler.Default);
+            // Only start the background processing task if foreground work is allowed
+            if (ForegroundKind != ForegroundThreadDataKind.Unknown)
```

Comment:
don't we have test foreground notification service?

http://source.roslyn.io/#Roslyn.Services.Test.Utilities/Workspaces/TestForegroundNotificationService.cs,12

Reply:
Yes we should remove that but it's still on the backlog.

Reply:
so. we no longer use mock on tests? is that it?

Reply:
Is it safe to silently discard these? Maybe a non-fatal error could be logged.

Reply:
cc @KirillOsenkov we used to have no logging for these notifications and we had them silently discarded. We need to verify we're doing the right thing here if this patch goes in.

Reply:
@Therzok I'm not sure what the specific concern is here. With this change, it is impossible to queue work for the background thread, hence why I didn't start the task.

Reply:
From what I see, the fallback thread data will not queue work. The concern is needing to know which things need to be setup to properly function.

We had code throwing in this service, that was masked, because we didn't have FatalError handlers setup.

Maybe it's worth notifying the consumer that the setup is not properly done?

As another point, I was thinking about filing a bug about FatalError not saying no handler is registered. Maybe ForehroundNotificationService should report that no thread handler is set?

Reply:
> From what I see, the fallback thread data will not queue work.

Correct. It will throw a synchronous exception immediately upon attempting to schedule the work.

Reply:
Ohhhh, I just noticed the `Contract.ThrowIfTrue` calls below. No worries from my side anymore!